sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala

```diff
@@ -506,18 +513,18 @@ case class RangeExec(range: org.apache.spark.sql.catalyst.plans.logical.Range)
       |       $numElementsTodo = 0;
       |       if ($nextBatchTodo == 0) break;
       |     }
-      |     $numOutput.add($nextBatchTodo);
-      |     $inputMetrics.incRecordsRead($nextBatchTodo);
       |     $batchEnd += $nextBatchTodo * ${step}L;
       |   }
       |
       |   int $localEnd = (int)(($batchEnd - $nextIndex) / ${step}L);
       |   for (int $localIdx = 0; $localIdx < $localEnd; $localIdx++) {
       |     long $value = ((long)$localIdx * ${step}L) + $nextIndex;
       |     ${consume(ctx, Seq(ev))}
-      |     $shouldStop
+      |     $stopCheck
       |   }
       |   $nextIndex = $batchEnd;
+      |   $numOutput.add($localEnd);
```

Comment:
I image a case. There is range + limit + a blocking op + ...

Now as at the range there is no `stopCheck`, right?

Assume a range batch is 1000. Because there is no `stopCheck`, this loop on `localIdx` will run to end. Although the limit works to only pass `n` rows into the blocking op, here we still add `localEnd` into `numOutput`.

I've not really tested it. Not sure if it is really a problem. Since it is late, I may check it more tomorrow if it has not figured out yet.

Reply:
that's expected isn't it? The range operator does output 1000 rows, the limit operator takes 1000 inputs, but only output like 100 rows.

Reply:
more background: the stop check for limit is done in batch granularity, while the stop check for result buffer is done in row granularity.

That said, even if the limit is smaller than the batch size, the range operator still outputs a entire batch, physically.

Reply:
If it is, then it is no problem. I was thinking that the number of output metric at range operator should be 100 if it is followed by a limit(100) operator.